src/Plugin/Commerce/PaymentGateway/Heartland.php

```diff
@@ -0,0 +1,391 @@
+            '#required' => true,
+        ];
+
+        // $form['store_card_data'] = [
+        //     '#type' => 'checkbox',
+        //     '#title' => $this->t('Store Card Data'),
+        //     '#description' => $this->t('Selet this option to store card data.  You must have multi-use token enabled on your Heartland Payment Systems account.'),
+        //     '#default_value' => FALSE,
+        // ];
+  
+        return $form;
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function validateConfigurationForm(array &$form, FormStateInterface $form_state)
+    {
+        parent::validateConfigurationForm($form, $form_state);
+
+        // Check if keys match the environment test/live
+        if (!$form_state->getErrors()) {
+            $values = $form_state->getValues($form['#parents'])['configuration']['commerce_heartland'];
+            $mode = $values['mode'];
+            $public_env = explode('_', $values['public_key'])[1];
+            $secret_env = explode('_', $values['secret_key'])[1];
+
+            if (($mode == 'test' && $public_env == 'prod') || ($mode == 'live' && $public_env == 'cert')) {
+                $form_state->setError($form['public_key'], $this->t('Your public key does not match the mode (@mode).', ['@mode' => $values['mode']]));
+            }
+
+            if (($mode == 'test' && $secret_env == 'prod') || ($mode == 'live' && $secret_env == 'cert')) {
+                $form_state->setError($form['secret_key'], $this->t('Your secret key does not match the mode (@mode).', ['@mode' => $values['mode']]));
+            }
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function submitConfigurationForm(array &$form, FormStateInterface $form_state)
+    {
+        parent::submitConfigurationForm($form, $form_state);
+  
+        if (!$form_state->getErrors()) {
+            $values = $form_state->getValue($form['#parents']);
+            $this->configuration['public_key'] = $values['public_key'];
+            $this->configuration['secret_key'] = $values['secret_key'];
+            // $this->configuration['store_card_data'] = $values['store_card_data'];
+        }
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function createPayment(PaymentInterface $payment, $capture = true)
+    {
+        $this->assertPaymentState($payment, ['new']);
+        $payment_method = $payment->getPaymentMethod();
+        $this->assertPaymentMethod($payment_method);
+
+        // Perform the create payment request here, throw an exception if it fails.
+        // See \Drupal\commerce_payment\Exception for the available exceptions.
+        // Remember to take into account $capture when performing the request.
+
+        // Set some variables to use for our transaction
+        $amount = $payment->getAmount();
+        $number = $amount->getNumber();
+        $currency = $amount->getCurrencyCode();
+        $remote_id = $payment_method_token = $payment_method->getRemoteId();
```

Comment:
Does `$remote_id` need to be an alias of `$payment_method_token` here? `$remote_id` doesn't seem to be used before it's set to the transaction ID after a successful transaction.

Reply:
No it doesn't.